paddle/fluid/imperative/reducer.cc

```diff
@@ -0,0 +1,348 @@
+
+namespace paddle {
+namespace imperative {
+
+#if defined(PADDLE_WITH_NCCL)
+std::shared_ptr<Reducer> Reducer::s_instance_ = NULL;
+
+Reducer::Reducer(const std::vector<std::shared_ptr<imperative::VarBase>> &vars,
+                 const std::vector<std::vector<size_t>> &group_indices,
+                 const std::vector<bool> &is_sparse_gradient,
+                 std::shared_ptr<imperative::ParallelContext> parallel_ctx)
+    : vars_(vars),
+      group_indices_(group_indices),
+      is_sparse_gradient_(is_sparse_gradient),
+      parallel_ctx_(parallel_ctx) {
+  VLOG(3) << "Start construct the Reducer ...";
+  // initialize groups
+  InitializeGroups(group_indices);
+
+  {
+    for (size_t group_index = 0; group_index < group_indices.size();
+         ++group_index) {
+      for (size_t var_index = 0; var_index < group_indices[group_index].size();
+           ++var_index) {
+        size_t global_var_index = group_indices[group_index][var_index];
+        const auto variable_index = VariableIndex{
+            .group_index = group_index, .inside_group_index = var_index,
+        };
+        VLOG(3) << "add hook for var[" << vars_[global_var_index]->GradVarName()
+                << "], it's in group [" << group_index << "]";
+        vars_[global_var_index]->SharedVar()->AddGradVarLeafBackwardHook(
+            std::unique_ptr<LambdaGradAccumulatorPostHook>(
+                new LambdaGradAccumulatorPostHook([=](VariableWrapper *grad) {
+                  this->AddDistHook(grad, variable_index);
+                })));
+      }
+    }
+  }
+
+  compute_stream_ = static_cast<platform::CUDADeviceContext *>(
+                        platform::DeviceContextPool::Instance().Get(place_))
+                        ->stream();
+  comm_stream_ = platform::NCCLCommContext::Instance().Get(0, place_)->stream();
+  events_.resize(group_indices.size());
+  for (auto &event : events_) {
+    event = platform::CudaEventResourcePool::Instance().New(
+        BOOST_GET_CONST(platform::CUDAPlace, place_).device);
+  }
+  comm_enent_ = platform::CudaEventResourcePool::Instance().New(
+      BOOST_GET_CONST(platform::CUDAPlace, place_).device);
+
+  std::call_once(once_flag_, []() {
+    std::atexit([]() { Reducer::GetInstance()->ReleaseReducer(); });
+  });
+}
+
+void Reducer::ReleaseReducer() {
+  for (auto &event : events_) {
+    event.reset();
+  }
+  comm_enent_.reset();
+}
+
+int64_t Reducer::InitializeDenseGroups(
+    const std::vector<size_t> &variable_indices_, Group *p_group) {
+  int64_t all_length = 0;
+  for (size_t index = 0; index < variable_indices_.size(); ++index) {
+    const auto variable_index = variable_indices_[index];
+    const auto &var = vars_[variable_index];
+    const auto var_name = var->Name();
+    PADDLE_ENFORCE_EQ(is_sparse_gradient_[variable_index], false,
+                      platform::errors::PreconditionNotMet(
+                          "Tensor `%s` 's GRAD must be LoDTensor.", var_name));
```

Comment:
the blank between `%s` and `'s` need to be removed? 
maybe appending `but received GRAD is SelectedRows or SparseTensor.` is better?